---
title: "The journey of Multiple Sequence Alignment"
subtitle: "Diving in the Notum"
author: "Constantinos Yeles (Konstantinos Geles)"
date: "First version: Tue 26/07/2021, Latest update: `r format(Sys.time(), '%a %d/%b/%Y')`"
output:
  html_document:
    toc: yes
    toc_depth: 3
    theme: paper 
  pdf_document:
    toc: yes
    toc_depth: 3
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

# Blast alignment
First we need to download the fasta sequence and the database from
dropbox? Maybe can be uploaded later to zenodo 
(probably I will have more context to add later). 

We will use local blast run on docker as the [link](https://faculty.virginia.edu/wrpearson/fasta/) wasn't working.

```{bash}
# get the docker and run
docker pull ncbi/blast
docker run --rm -ti -v $(pwd):/home/my_data  ncbi/blast
# move inside /home dir of the container
cd /home

# create a folder for the ouput of the blast database and unzip the fasta sequences
mkdir my_data/some_species_blast_db
gunzip my_data/oma_db.gz 

# make a blast database 
makeblastdb  -dbtype "prot" -in "my_data/oma_db" -input_type "fasta" -out "my_data/some_species_blast_db/blast_oma.db" -logfile "my_data/some_species_blast_db/logfile"

# blast the sequence of Notum
blastp -query "my_data/dm_notum.fasta"  -db "my_data/some_species_blast_db/blast_oma.db" -out "my_data/results_blast.txt" -outfmt "7 delim=@ qseqid qlen sseqid slen qstart qend sstart send qseq sseq evalue bitscore score length pident nident mismatch positive ppos frames qframe sframe sstrand qcovs qcovhsp qcovus" -num_threads 4
```

# Import results into R
```{r include=FALSE, eval = TRUE}
# load libraries
suppressPackageStartupMessages({
library(tidyverse)
library(plyranges)
library(vroom)
  })

# function to import the results
read_blast_tab_format <- function(blast_file, delim = "@"){
  fourth_line_fields <- blast_file %>% 
  vroom_lines() %>% 
  .[4] %>% 
  str_remove("# Fields: ") %>% 
  str_split(", ") %>% 
  unlist %>% 
  str_replace_all(" |/","_") %>% 
  str_replace("%","percent") %>% 
  str_remove("\\.")
  
  blast_sequences <- blast_file %>% 
  read_delim(delim = delim, # here it should be made clear that the delim could be something else
             comment = "#", 
             col_names = fourth_line_fields) %>% 
  mutate(strand = ifelse(subject_strand == "minus","-","+") %>% as_factor()) %>% 
  mutate(
         start = ifelse(strand == "-", yes = s_end, no = s_start),
         end = ifelse(strand == "-", yes = s_start,no = s_end),
         seqnames = subject_id) %>% 
    as_granges()
  
  return(blast_sequences)
}
# import the results

notum_blst <- read_blast_tab_format(blast_file = "results_blast.txt")
```

## print an example of the results 
```{r eval = TRUE}
notum_blst %>% arrange(evalue)
```

## get the 10 top hits identifiers
```{r eval = TRUE}
top10_seqs <- notum_blst %>% arrange(evalue) %>% head(11) %>% .$subject_id
```

## import fasta file with all sequences
import the sequences and filter them for the 10 best hits from the previous results in order to make the multiple sequence alignments
```{r eval = TRUE}
suppressPackageStartupMessages({
library(Biostrings)
  })
fasta_db <- readAAStringSet("oma_db")
fasta_db <- fasta_db[sort(names(fasta_db))]

# I didn't consider that the names will be cut in the db creation thus I have to manipulate them to work properly for the filtering
names_fst_db <- fasta_db  %>% names %>% str_remove(" .+")
 
# filter for the top 10
fasta_db_fil <- fasta_db[names_fst_db %in% top10_seqs]
```

# Multiple Sequence Alignment
I found two packages in R implementing various algorithms. 
One is [msa](https://www.bioconductor.org/packages/release/bioc/html/msa.html) and the other [DECIPHER](https://www.bioconductor.org/packages/release/bioc/html/DECIPHER.html)
for now I will try msa
```{r eval = TRUE}
suppressPackageStartupMessages({
library(msa)
library(bios2mds)
  })
myFirstAlignment <- msa(fasta_db_fil)
```

## print the MSA
```{r eval = TRUE}
print(myFirstAlignment, show="complete")
```

## get significant sequences
```{r eval = TRUE}
sign_seqs <- notum_blst %>% filter(evalue < 0.05) %>% .$subject_id

# filter for the significant sequences
fasta_db_fil <- fasta_db[names_fst_db %in% sign_seqs]
```

## perform MSA with MUSCLE
```{r}
sign_seqs_muscle <- msaMuscle(fasta_db_fil)
sign_seqs_bio2mds <- msa::msaConvert(sign_seqs_muscle, type = "bios2mds::align")
export.fasta(sign_seqs_bio2mds, outfile = "sign_seqs_msa_muscle.fa", ncol = 60, open = "w")
```
check the sign_seqs_msa_muscle.fa file for the alignments

# IQTREE
```{bash}
docker run --rm -ti -v $(pwd):/home/my_data staphb/iqtree

cd home
mkdir my_data/my_first_tree

iqtree -s my_data/sign_seqs_msa_muscle.fa -nt 6 -mset LG -pre my_data/my_first_tree/my_tree

```

# Following the "Estimating phylogenetic trees with phangorn"
```{r}
library(ape)
library(phangorn)
```
## Distance based methods

### UPGMA
```{r}
sign_seqs_phyDat <- as.phyDat(sign_seqs_muscle, type = "AA")

dm  <- dist.ml(sign_seqs_phyDat)

treeUPGMA  <- upgma(dm)

treeNJ  <- NJ(dm)

plot(treeUPGMA, main="UPGMA")

plot(treeNJ, "unrooted", main="NJ")
```

### Bootstrap
```{r}
fun <- function(x) upgma(dist.ml(x))

bs_upgma <- bootstrap.phyDat(sign_seqs_phyDat,  fun)

#bs_upgma <- bootstrap.phyDat(sign_seqs_phyDat,  \(x){dist.ml(x) |> upgma})

plotBS(treeUPGMA, bs_upgma, main="UPGMA")
```

### Parsimony
```{r}
parsimony(treeUPGMA, sign_seqs_phyDat)
parsimony(treeNJ, sign_seqs_phyDat)

treePars  <- optim.parsimony(treeUPGMA, sign_seqs_phyDat)

#parsimony ratchet (Nixon 1999)
treeRatchet  <- pratchet(sign_seqs_phyDat, trace = 0)
parsimony(c(treePars, treeRatchet), sign_seqs_phyDat)

treeRatchet  <- acctran(treeRatchet, sign_seqs_phyDat)

plotBS(midpoint(treeRatchet), type="phylogram")
```

### Maximum likelihood
```{r}
fit = pml(treeNJ, data=sign_seqs_phyDat)
fit
fitJC  <- optim.pml(fit, TRUE)
logLik(fitJC)

fitWAG <- update(fit, k=4, inv=0.2)

fitWAG <- optim.pml(fitWAG, model="Dayhoff_DCMut", optInv=TRUE, optGamma=TRUE,
    rearrangement = "stochastic", control = pml.control(trace = 0))
fitGTR
```

# HMMER
## hmmsearch
```{bash}
docker run --rm -ti -v $(pwd):/home/my_data  staphb/hmmer
cd ..

mkdir my_data/hmm_search

hmmsearch --cut_ga --domtblout my_data/hmm_search/PAE_domtbl.out my_data/PAE.hmm my_data/oma_db > /dev/null
```

## Import hmmsearch results into R
```{r}
hmmsearch_result <- "hmm_search/PAE_domtbl.out"
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target")

notum_hmmsearch <- data.table::fread(cmd = "grep -v '^#' hmm_search/PAE_domtbl.out", 
                  header = FALSE, sep = " ",  fill=TRUE) %>% 
  unite("V23",c(str_c("V",23:33)))

names(notum_hmmsearch) <- second_line_fields
```

## Multiple Sequence Alignment
```{r}
# filter for the hmmsearch complete sequences
fasta_db_fil <- fasta_db[names_fst_db %in% notum_hmmsearch$target_name]

# filter for the hmmsearch env_coord sequences
fasta_db_fil_with_len <- fasta_db[names_fst_db %in% notum_hmmsearch$target_name]

fasta_db_fil_with_len <- fasta_db_fil_with_len[sort(names(fasta_db_fil_with_len))]

notum_hmmsearch_clm <- notum_hmmsearch %>% 
  filter(!duplicated(target_name)) %>% 
  select(target_name, from_env_coord, to_env_coord) %>% 
  arrange(target_name)

#fasta_db_fil_with_len %>% head %>% Biostrings::as.list() %>% Biostrings::AAStringSet()
fasta_db_fil_with_len <- pmap(list(fasta_db_fil_with_len %>% as.list, 
          notum_hmmsearch_clm %>% select(target_name, from_env_coord) %>% deframe %>% as.list(),
          notum_hmmsearch_clm %>% select(target_name, to_env_coord) %>% deframe %>% as.list()),
     subseq) %>% Biostrings::AAStringSet()
```

## perform MSA with MUSCLE hmmsearch
```{r}
hmmsearch_muscle <- msaMuscle(fasta_db_fil)
hmmsearch_bio2mds <- msa::msaConvert(hmmsearch_muscle, type = "bios2mds::align")
export.fasta(hmmsearch_bio2mds, outfile = "hmmsearch_msa_muscle.fa", ncol = 60, open = "w")

## env_coord
hmmsearch_env_coord_muscle <- msaMuscle(fasta_db_fil_with_len)
hmmsearch_env_coord_bio2mds <- msa::msaConvert(hmmsearch_env_coord_muscle, type = "bios2mds::align")
export.fasta(hmmsearch_env_coord_bio2mds, outfile = "hmmsearch_env_coord_msa_muscle.fa", ncol = 60, open = "w")
```
check the sign_seqs_msa_muscle.fa file for the alignments

## IQTREE hmmsearch
```{bash}
docker run --rm -ti -v $(pwd):/home/my_data staphb/iqtree

cd home

iqtree -s my_data/hmmsearch_msa_muscle.fa -nt 8 -mset LG -pre my_data/my_hmmsearch_tree/my_hmmsearch_tree

iqtree -s my_data/hmmsearch_env_coord_msa_muscle.fa -nt 8 -mset LG -pre my_data/my_hmmsearch_tree/my_hmmsearch_env_coord_tree

```

# wnt Family
## HMMER:hmmsearch
```{bash}
docker run --rm -ti -v $(pwd):/home/my_data  staphb/hmmer
cd ..
cd home

hmmsearch --cut_ga --domtblout my_data/hmm_search/wnt_domtbl.out my_data/wnt.hmm my_data/oma_db > /dev/null
```

## Import hmmsearch results into R
```{r}
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target")

wnt_hmmsearch <- data.table::fread(cmd = "grep -v '^#' hmm_search/wnt_domtbl.out | awk -v OFS='\t'  '{for(i=1;i<=23;i++) printf $i\"\t\"; print \"\"}'",  header = FALSE, sep = "\t",  fill=TRUE)

names(wnt_hmmsearch)[1:23] <- second_line_fields
```

## Multiple Sequence Alignment
```{r}
# filter for the hmmsearch env_coord sequences
fasta_db_fil_with_len <- fasta_db[names_fst_db %in% wnt_hmmsearch$target_name]

fasta_db_fil_with_len <- fasta_db_fil_with_len[sort(names(fasta_db_fil_with_len))]

wnt_hmmsearch <- wnt_hmmsearch %>% 
  filter(!duplicated(target_name)) %>% 
  select(target_name, from_env_coord, to_env_coord) %>% 
  arrange(target_name)

#fasta_db_fil_with_len %>% head %>% Biostrings::as.list() %>% Biostrings::AAStringSet()
fasta_db_fil_with_len <- pmap(list(fasta_db_fil_with_len %>% as.list, 
          wnt_hmmsearch %>% select(target_name, from_env_coord) %>% deframe %>% as.list(),
          wnt_hmmsearch %>% select(target_name, to_env_coord) %>% deframe %>% as.list()),
     subseq) %>% Biostrings::AAStringSet()
```

## perform MSA with MUSCLE hmmsearch
```{r}
## env_coord
hmmsearch_env_coord_muscle <- msaMuscle(fasta_db_fil_with_len)
hmmsearch_env_coord_bio2mds <- msa::msaConvert(hmmsearch_env_coord_muscle, type = "bios2mds::align")
export.fasta(hmmsearch_env_coord_bio2mds, outfile = "wnt_hmmsearch_env_coord_msa_muscle.fa", ncol = 60, open = "w")
```
check the sign_seqs_msa_muscle.fa file for the alignments

## IQTREE hmmsearch
```{bash}
docker run --rm -ti -v $(pwd):/home/my_data staphb/iqtree

cd home

iqtree -s my_data/wnt_hmmsearch_env_coord_msa_muscle.fa -nt 8 -mset LG -pre my_data/my_hmmsearch_tree/wnt_hmmsearch_env_coord_tree

```

# Searching for the Hs PORCN, protein-cysteine N-palmitoyltransferase porcupine isoform X5 [Homo sapiens]
## the Fasta file
```{bash}
cat porcn_hs.fasta 
>Hs|PORCN  gi|530421517|ref|XP_005272694.1| gi|530421517|ref|XP_005272694.1| PREDICTED: protein-cysteine N-palmitoyltransferase porcupine isoform X5 [Homo sapiens]
MYAIPSPSDWTVHSTGETEVLHSLQVTGCACACRTPTCTQLVAQSDVPDFSRLPPVLQYR
ASTWAGQPQTEISIHLAIHPWGSAMATFSRQEFFQQLLQGCLLPTAQQGLDQIWLLLAIC
LACRLLWRLGLPSYLKHASTVAGGFFSLYHFFQLHMVWVVLLSLLCYLVLFLCRHSSHRG
VFLSVTILIYLLMGEMHMVDTVTWHKMRGAQMIVAMKAVSLGFDLDRGEVGTVPSPVEFM
GYLYFVGTIVFGPWISFHSYLQAVQGRPLSCRWLQKVARSLALALLCLVLSTCVGPYLFP
YFIPLNGDRLLRNKKRKARGTMVRWLRAYESAVSFHFSNYFVGFLSEATATLAGAGFTEE
KDHLEWDLTVSKPLNVELPRSMVEVVTSWNLPMSYWLNNYVFKNALRLGTFSAVLVTYAA
SALLHGFSFHLAAVLLSLAFITYVEHVLRKRLARILSACVLSKRCPPDCSHQHRLGLGVR
ALNLLFGALAIFHLAYLGSLFDVDVDDTTEEQGYGMAYTVHKWSELSWASHWVTFGCWIF
YRLIG
```

## PFAM search
we search the sequence on the PFAM site and we get the MBOAT, 
membrane-bound O-acyltransferase family as result.
We download the model for the MBOAT

## HMMER:hmmsearch
```{bash}
docker run --rm -ti -v $(pwd):/home/my_data  staphb/hmmer

cd /home

hmmsearch --cut_ga --domtblout my_data/hmm_search/MBOAT_domtbl.out my_data/MBOAT.hmm my_data/oma_db > /dev/null
```

## Import hmmsearch results into R
```{r}
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target")

MBOAT_hmmsearch <- data.table::fread(cmd = "grep -v '^#' hmm_search/MBOAT_domtbl.out | awk -v OFS='\t'  '{for(i=1;i<=23;i++) printf $i\"\t\"; print \"\"}'",  header = FALSE, sep = "\t",  fill=TRUE)

names(MBOAT_hmmsearch)[1:23] <- second_line_fields
```

## Multiple Sequence Alignment
```{r}
# filter for the hmmsearch env_coord sequences
fasta_db_fil_with_len <- fasta_db[names_fst_db %in% MBOAT_hmmsearch$target_name]

MBOAT_hmmsearch <- MBOAT_hmmsearch %>% 
  filter(!duplicated(target_name)) %>% 
  select(target_name, from_env_coord, to_env_coord) %>% 
  arrange(target_name)

identical(fasta_db_fil_with_len %>% names %>% str_remove(" .+"),
          MBOAT_hmmsearch$target_name)

#fasta_db_fil_with_len %>% head %>% Biostrings::as.list() %>% Biostrings::AAStringSet()
fasta_db_fil_with_len <- pmap(list(fasta_db_fil_with_len %>% as.list, 
          MBOAT_hmmsearch %>% select(target_name, from_env_coord) %>% deframe %>% as.list(),
          MBOAT_hmmsearch %>% select(target_name, to_env_coord) %>% deframe %>% as.list()),
     subseq) %>% Biostrings::AAStringSet()
```

## perform MSA with MUSCLE hmmsearch
```{r}
## env_coord
hmmsearch_env_coord_muscle <- msaMuscle(fasta_db_fil_with_len, type = "protein")
hmmsearch_env_coord_bio2mds <- msa::msaConvert(hmmsearch_env_coord_muscle, type = "bios2mds::align")
export.fasta(hmmsearch_env_coord_bio2mds, outfile = "MBOAT_hmmsearch_env_coord_msa_muscle.fa", ncol = 60, open = "w")
```
check the sign_seqs_msa_muscle.fa file for the alignments

## IQTREE hmmsearch
```{bash}
docker run --rm -ti -v $(pwd):/home/my_data staphb/iqtree

cd home

iqtree -s my_data/MBOAT_hmmsearch_env_coord_msa_muscle.fa -nt 4 -mset LG -pre my_data/my_hmmsearch_tree/MBOAT_hmmsearch_env_coord_tree

```

# hmmbuild
```{bash}
hmmbuild --amino --cpu 4  my_data/NOTUM_try my_data/hmmsearch_env_coord_msa_muscle.fa 
hmmbuild --amino --cpu 4  my_data/WIF_try my_data/WIF_hmmsearch_env_coord_msa_muscle.fa 
hmmbuild --amino --cpu 4  my_data/wnt_try my_data/wnt_hmmsearch_env_coord_msa_muscle.fa 
```

# hmmscan WIF proteins in pfam domains
## export the WIF sequences from oma.db
```{r}
fasta_db_fil_with_len %>% Biostrings::writeXStringSet("WIF_oma_seq.fasta")
```


```{bash}
hmmpress my_data/pfam/Pfam-A.hmm

hmmscan --cut_ga --domtblout my_data/hmm_search/WIF_profile_domtbl.out my_data/pfam/Pfam-A.hmm my_data/WIF_oma_seq.fasta > /dev/null

```
## Import hmmscan results into R
```{r}
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target_1",
                        "description_of_target_2","description_of_target_3")

WIF_hmmscan <- data.table::fread(cmd = "grep -v '^#' hmm_search/WIF_profile_domtbl.out | awk -v OFS='\t'  '{for(i=1;i<=29;i++) printf $i\"\t\"; print \"\"}'",  header = FALSE, sep = "\t",  fill=TRUE)

names(WIF_hmmscan)[1:25] <- second_line_fields

WIF_hmmscan %>% 
  distinct(target_name, query_name) %>% 
  pivot_wider(names_from = target_name, values_from = target_name,values_fill = "0") %>% 
  unite(col = "domains",2:11) %>% 
  mutate(domains = domains %>% str_remove_all("0_") %>% str_remove_all("_0"),
         query_name = query_name %>% str_replace("\\|","_")) %>% 
  unite(query_name, 1:2, sep = "__DOMAIN__") %>% view()
```

# Search four "representative" proteins that have the WIF domain respect to sponges
In the tree of WIF from oma.db we can see 4 big clusters that have particular characteristics
One is the cluster of the Human WIF (HS|WIF1) the the human with RYK domain (Hs|RYK) then those with 
NTR domain and then those with the "pure"  PK_TYR_SER_THR_PKinase
We will tblastn from each cluster one and evaluate some of the results
## The selected sequences:

### Hs|WIF1 
```{bash}
>Hs|WIF1 gi|111125011|ref|NP_009122.2| gi|111125011|ref|NP_009122.2| wnt inhibitory factor 1 precursor [Homo sapiens]
MARRSAFPAAALWLWSILLCLLALRAEAGPPQEESLYLWIDAHQARVLIGFEEDILIVSEGKMAPFTHDFRKAQQRMPAI
PVNIHSMNFTWQAAGQAEYFYEFLSLRSLDKGIMADPTVNVPLLGTVPHKASVVQVGFPCLGKQDGVAAFEVDVIVMNSE
GNTILQTPQNAIFFKTCQQAECPGGCRNGGFCNERRICECPDGFHGPHCEKALCTPRCMNGGLCVTPGFCICPPGFYGVN
CDKANCSTTCFNGGTCFYPGKCICPPGLEGEQCEISKCPQPCRNGGKCIGKSKCKCSKGYQGDLCSKPVCEPGCGAHGTC
HEPNKCQCQEGWHGRHCNKRYEASLIHALRPAGAQLRQHTPSLKKAEERRDPPESNYIW
```

### Hs|RYK (picle Rick)
```{bash}
>Hs|RYK gi|54607020|ref|NP_001005861.1| gi|54607020|ref|NP_001005861.1| tyrosine-protein kinase RYK isoform 1 precursor [Homo sapiens]
MRGAARLGRPGRSCLPGARGLRAPPPPPLLLLLALLPLLPAPGAAAAPAPRPPELQSASAGPSVSLYLSEDEVRRLIGLD
AELYYVRNDLISHYALSFSLLVPSETNFLHFTWHAKSKVEYKLGFQVDNVLAMDMPQVNISVQGEVPRTLSVFRVELSCT
GKVDSEVMILMQLNLTVNSSKNFTVLNFKRRKMCYKKLEEVKTSALDKNTSRTIYDPVHAAPTTSTRVFYISVGVCCAVI
FLVAIILAVLHLHSMKRIELDDSISASSSSQGLSQPSTQTTQYLRADTPNNATPITSSLGYPTLRIEKNDLRSVTLLEAK
GKVKDIAISRERITLKDVLQEGTFGRIFHGILIDEKDPNKEKQAFVKTVKDQASEIQVTMMLTESCKLRGLHHRNLLPIT
HVCIEEGEKPMVILPYMNWGNLKLFLRQCKLVEANNPQAISQQDLVHMAIQIACGMSYLARREVIHKDLAARNCVIDDTL
QVKITDNALSRDLFPMDYHCLGDNENRPVRWMALESLVNNEFSSASDVWAFGVTLWELMTLGQTPYVDIDPFEMAAYLKD
GYRIAQPINCPDELFAVMACCWALDPEERPKFQQLVQCLTEFHAALGAYV
```

### Of|1176102141
```{bash}
>Of|1176102141 XP_020621816.1| uncharacterized protein LOC110059462 [Orbicella faveolata]
MAGLEARNFLKVLFLMSVPVLGKCSTKLNVFLDAREVSRFLYDLNTNLFLVQDGVVAPILKTPILSSMIPPIEPYIKDLK
FKFNATGMVQYSLNFKSSNETIMKHPTANIRLNGLVPRKTKVFRVKFPCTGKVIGGVLLTINISFVDVAGSEIWGPLSLT
LKRQCVSGNDDLINRYNSTESGEVSPLPAPQICNKRCSKRHVMRRFCLSDFVIKARMESEVMRNGLPRLRLRIARTYKQG
KVKINNKNQLMEKRGQEITCSCSNLKVNKVYLILGKEDKRKRVLFLDNFSTALEWNKNGKSYVKTFRNRSSCPERKMP
```

### Ci|198430807
```{bash}
>Ci|198430807 ref|XP_002129769.1| PREDICTED: tyrosine-protein kinase RYK [Ciona intestinalis]
MRVCAYILAVVLLPCMSKCSIDLYLAEDQVEKLFGIADAEMYYIRNGIINENAVNFVAHVQPEHTDLTYLWKVTNKAITV
AYDLDILDVQKSPAQLNTGSVEMNKHSILYPPKLESKSGTIRYSEDLQPQTLKVSLVCTGEREGNVTVTMKLKLVMDSGT
DLVVNIKRLKKCLKTHSFVEDKPVLTDQPSEPNDQDLNLGGTQQSTSAFYISVSVVSSFIVLIVVGVTLWHVRMSKRYQR
SGVDNCIPMQPGAQHLNQFLRPDLPNNALKPPSHSSLMQIITPLINDLGTDINEIQSKLSAIAIPRENVDIGELTLKGTF
ARIYKGTLDDGTPVLIKTVSDMATEEQKRLLLFESSLLRGMHHRNLLSMKHVVLQDDSTPPMVLFHFMEGGNLKHYLQGL
KLKESPGLVSGIFGSGSTKYEQVSTQDLVEMAIQIACGMTYLSKRGLVHKDLAARNCVIDRDLKVKISDNALSRDAFPAD
YCCLGDNENRPVRWLGLESLVHKVYSSASDVWSFGVLLWELQSLAATPYADIDDFEMASYLKDGFRLSQPINCPDHLYSI
IASCWHYSPDQRPSFTKLLQMLTEFYAELEVYV
```
## Results:
### Hs|WIF1 vs Halisarca caerulea comp25790_c0_seq1 transcribed RNA sequence
```{bash}
>Hc|comp25790
CQRAECIPGCLNGGTCAGPNWCRCARGWSGNRCRRAVCIPACNNGGKCTAPNTCECAEGWTGNQCQRPVCDSECLNGGIC
AAPNTCQCSSGWSGDHCDLPVCRPKCTNGGICIRPSKCLCDSSWTGKRCQKPVCRPGCENGGTCTEPNTCQCTSGWSGDR
C
```
### Hs|WIF1 vs Halichondria panicea TRINITY_DN792_c2_g1_i4, transcribed RNA sequence
```{bash}
>Hp|TRINITY_DN792_c2_g1_i4
CENGGICTQERTCHCAPGYEGPLCEEAVCFPPCTNGGNCTAPGFCECPREWEGIRCEDPGCSPSCENGGTCIRPGECYCT
PQWTGDNCQDAICNPPCGNGGHCISPGHCECGGTGFLGSRCTEAVCKPACENGGECVEPNVCQCLSLWTGNYCDT
```
### Hs|WIF1 vs Aplysina aerophoba TR173969_c0_g2_i1 transcribed RNA sequence
```{bash}
>Ae|TR173969_c0_g2_i1
RRCRKAICNEPCRNGGTCTAPNVCACADGYMGQLCEVPVCSRECQNGGICIGKDECQCEMGYSGLHCESAICTDSCMNGG
TCVEPETCTCPANYTGRVCSEPVCIRPCLNVGQCTGPGVCTCTAEFTGDLCEVPICLDPCLNGGYCSAPNVCTCEDGYEG
HNC
```
### Hs|WIF1 vs Leucetta chagosensis breed sponge isolate calcarea TRINITY_DN9568_c1_g1_i1, transcribed RNA sequence
```{bash}
>Lc|TRINITY_DN9568_c1_g1_i1
PKSAWTGSRCQTPVCLPSCQNGGQCVSPQTCSCAAGYTGSSCGNAVCSTSCKNGGSCVSPNKCSCPAGYTGSYCEKAVCS
TSCRNGGSCVRPNTCSCPAGYTGSYCEKAVCSTSCQNGGSCVRPNTCSCPAGYTGSYCEKAVCSTSCKNGGSCVSPNTCS
CPAGYTGSYCEK
```

```{bash}
hmmpress my_data/pfam/Pfam-A.hmm

hmmscan --cut_ga --domtblout my_data/hmm_search/WIF_profile_HS_WIF1vssponges_domtbl.out my_data/pfam/Pfam-A.hmm my_data/Hs_WIF1vs_sponges.fasta > /dev/null
```

## Import hmmscan of WIF1 vs sponges results into R
```{r}
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target_1",
                        "description_of_target_2","description_of_target_3")

WIF_hmmscan <- data.table::fread(cmd = "grep -v '^#' hmm_search/WIF_profile_HS_WIF1vssponges_domtbl.out | awk -v OFS='\t'  '{for(i=1;i<=29;i++) printf $i\"\t\"; print \"\"}'",  header = FALSE, sep = "\t",  fill=TRUE)

names(WIF_hmmscan)[1:25] <- second_line_fields

WIF_hmmscan %>% 
  distinct(target_name, query_name) %>% 
  pivot_wider(names_from = target_name, values_from = target_name,values_fill = "0") %>% 
  unite(col = "domains",2:11) %>% 
  mutate(domains = domains %>% str_remove_all("0_") %>% str_remove_all("_0"),
         query_name = query_name %>% str_replace("\\|","_")) %>% 
  unite(query_name, 1:2, sep = "__DOMAIN__") %>% view()
```


### Hs|RYK vs Halichondria panicea TRINITY_DN396839_c0_g1_i1, transcribed RNA sequence
```{bash}
>Hp|TRINITY_DN396839_c0_g1_i1
ESCLMKGFCHQNVNPILGACMENEKQPLLVYAYSTEGNLKKFLQRCKMSEMASSYALSTQQLVYIAIQIIRGVQYLHRKR
VVHRDVATRNCVIDHQLNVKVTDTALSRDLFPNDYHCLGDNENRPVKWLAIEALLDRHFTPASDVWSFGVTMWELMTLGQ
QPYADVDPFEMATYLKEGYRIAQPINCPDELFSVMACCWSMLGEERPKLTQLQACLQDFYTALGRFI
```

### Hs|RYK vs Cymbastela stipitata TRINITY_DN14054_c0_g1_i1, transcribed RNA sequence
```{bash}
>Cs|TRINITY_DN14054_c0_g1_i1
NIASVGAVATTDFSISFPCTNTLQGRADVRVTLNFTFDFTGGTLRSLVGQVFTMTGTRVCNCTNSRCDNTEPPNDGGVNN
EIFYIVIGCVGGVILLVIMVTVLYHCIMLGMWVKRRSQDEEPLPNIEDVIPMPPYPMVTITPLSSQVPIAPLPNSPRNSF
SVPSDLSKSLRSLQSFKDLIVDRKRISIGQVLKEGVFGVMYDGYLSNAEDDVEGAIPVIIKTVKDNTPDIVVKSLLEGGG
ATRAVSHRHVLPLLAAHVSDTEQPMLLFPKTSLGTLKGLLMRCREPKRGVSLTLTTQNLVFIAGQISRGMYHLTRKGLIH
RDLAARNIYVHENLHVKIGDRGLSWDLFPSEYSRLPDGEMCPIRWMSAEVLSERKYSHYSDVWSFGVVLWELMTLGRIPY
EEVPAEDMLAMLTAGHRLSQPKNCPDELFVLMGWCWALTPTDRPRFSHLTLRLKEFHEKINAFV
```

### Hs|RYK vs Neopetrosia compacta breed sponge isolate demosponge TRINITY_DN1516_c3_g1_i2, transcribed RNA sequence
```{bash}
>Nc|TRINITY_DN1516_c3_g1_i2
RDVFVDRKRIQFGCKITEGVFGVMYDGYLTNAEDDVEGAVPVIIKTVKDNTPDVVVKSLLEGGSATRGVSHRHLLPLLAA
HASDLEQPMLLFPKTAHGSLKSLLLCSRGAQRVIPGSPFLTTQDLVFMAEQVARGMYHLTRKGLVHKDLAARNIYVHENL
HVKIGDRGLCWDFFPSEYSTLEGDQELIASPIRWMAAEVIAEKKYSHYSDVWAFGVVLWEIMTKGKIPYEDVLPEQMYSY
LTYGSRLQQPKNCPDDLFSLMGWCWALTPGDRPRFSHLTIRLKEFHERLGAFI
```

### Hs|RYK vs Dysidea avara TR158463_c0_g1_i1 transcribed RNA sequence
```{bash}
>Da|TR158463_c0_g1_i1
MKDILVDKNRIQLGELIKEGTYGRMYEGYLANTEDDVEGSVSIIVKTVSDITPDQVVASLVDGCLVLRHIQHRHILQALG
VYSNEQEPVMVLLARTTFGTLKQFLDDIRLQGGPQSSYGKLLLTQDMVFISAQIARGMYHLSRKGGIIHKDLAARNIYIH
ENLHVKIGDRGLSWDFYPDDYQVTPNGEYVPVKWASVEVLEEQQYSMYSDVWAFGVVLWEIMTVGLVPYDSVTPNELVDY
LKSGQRLTRPTNCPDDMFSIMGQCWALTPTDRPKFSHLTVGLDHFHLRLNSYI
```

```{bash}
hmmpress my_data/pfam/Pfam-A.hmm

hmmscan --cut_ga --domtblout my_data/hmm_search/WIF_profile_HS_RYKvssponges_domtbl.out my_data/pfam/Pfam-A.hmm my_data/Hs_RYKvs_sponges.fasta > /dev/null
```

## Import hmmscan of WIF1 vs sponges results into R
```{r}
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target_1",
                        "description_of_target_2","description_of_target_3")

WIF_hmmscan <- data.table::fread(cmd = "grep -v '^#' hmm_search/WIF_profile_HS_RYKvssponges_domtbl.out | awk -v OFS='\t'  '{for(i=1;i<=29;i++) printf $i\"\t\"; print \"\"}'",  header = FALSE, sep = "\t",  fill=TRUE)

names(WIF_hmmscan)[1:25] <- second_line_fields

WIF_hmmscan %>% 
  distinct(target_name, query_name) %>% 
  pivot_wider(names_from = target_name, values_from = target_name,values_fill = "0") %>% 
  unite(col = "domains",2:11) %>% 
  mutate(domains = domains %>% str_remove_all("0_") %>% str_remove_all("_0"),
         query_name = query_name %>% str_replace("\\|","_")) %>% 
  unite(query_name, 1:2, sep = "__DOMAIN__") %>% view()
```

About Of|1176102141 we got only one result:
### Of|1176102141 vs Cymbastela stipitata TRINITY_DN17682_c0_g1_i1, transcribed RNA sequence
```{bash}
>Cs|TRINITY_DN17682_c0_g1_i1
LPPLPSSVDEISISWHASEEVIYELHVGQDPQGVFEDPRLNIPENGTIPRNKTAFSISFPC
```

```{bash}
touch my_data/Of1176102141_vs_sponges.fasta

hmmscan --cut_ga --domtblout my_data/hmm_search/WIF_profile_Of1176102141vssponges_domtbl.out my_data/pfam/Pfam-A.hmm my_data/Of1176102141_vs_sponges.fasta > /dev/null
```

## Import hmmscan of WIF1 vs sponges results into R
```{r}
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target_1",
                        "description_of_target_2","description_of_target_3")

WIF_hmmscan <- data.table::fread(cmd = "grep -v '^#' hmm_search/WIF_profile_Of1176102141vssponges_domtbl.out | awk -v OFS='\t'  '{for(i=1;i<=29;i++) printf $i\"\t\"; print \"\"}'",  header = FALSE, sep = "\t",  fill=TRUE)

names(WIF_hmmscan)[1:25] <- second_line_fields

WIF_hmmscan %>% 
  distinct(target_name, query_name) %>% 
  pivot_wider(names_from = target_name, values_from = target_name,values_fill = "0") %>% 
  unite(col = "domains",2:11) %>% 
  mutate(domains = domains %>% str_remove_all("0_") %>% str_remove_all("_0"),
         query_name = query_name %>% str_replace("\\|","_")) %>% 
  unite(query_name, 1:2, sep = "__DOMAIN__") %>% view()
```

### Ci|198430807 vs Halichondria panicea TRINITY_DN396839_c0_g1_i1, transcribed RNA sequence
```{bash}
>Hp|TRINITY_DN396839_c0_g1_i1
ESCLMKGFCHQNVNPILGACMENEKQPLLVYAYSTEGNLKKFLQRCKMSEMASSYALSTQQLVYIAIQIIRGVQYLHRKR
VVHRDVATRNCVIDHQLNVKVTDTALSRDLFPNDYHCLGDNENRPVKWLAIEALLDRHFTPASDVWSFGVTMWELMTLGQ
QPYADVDPFEMATYLKEGYRIAQPINCPDELFSVMACCWSMLGEERPKLTQLQACLQDFYTALGRFI
```

### Ci|198430807 vs Cymbastela stipitata TRINITY_DN14054_c0_g1_i1, transcribed RNA sequence
```{bash}
>Cs|TRINITY_DN14054_c0_g1_i1
LCLRAVWVLLALYRCVCALDWFIPAAEVQRIEGEENAPALYLVRQSVVQRSTLPLSETAMNTRVMFSWAGSGLAYQFKSV
SSNPAVLMGAEFENIASVGAVATTDFSISFPCTNTLQGRADVRVTLNFTFDFTGGTLRSLVGQVFTMTGTRVCNCTNSRC
DNTEPPNDGGVNNEIFYIVIGCVGGVILLVIMVTVLYHCIMLGMWVKRRSQDEEPLPNIEDVIPMPPYPMVTITPLSSQV
PIAPLPNSPRNSFSVPSDLSKSLRSLQSFKDLIVDRKRISIGQVLKEGVFGVMYDGYLSNAEDDVEGAIPVIIKTVKDNT
PDIVVKSLLEGGGATRAVSHRHVLPLLAAHVSDTEQPMLLFPKTSLGTLKGLLMRCREPKRGVSLTLTTQNLVFIAGQIS
RGMYHLTRKGLIHRDLAARNIYVHENLHVKIGDRGLSWDLFPSEYSRLPDGEMCPIRWMSAEVLSERKYSHYSDVWSFGV
VLWELMTLGRIPYEEVPAEDMLAMLTAGHRLSQPKNCPDELFVLMGWCWALTPTDRPRFSHLTLRLKEFHEKINAFV
```

### Ci|198430807 vs Halichondria panicea TRINITY_DN874_c0_g1_i3, transcribed RNA sequence
```{bash}
>Hp|TRINITY_DN874_c0_g1_i3
LTTDMPVSDLKSSPRHSHSSISVPPALNNSLRNLHSFRDLFVHRKRISIGVVLTEGVFGVIYDGYLSNAEDDVEGVTPVI
IKTVKDNTPEIVVRSLLEGGAALRHVPHRHLLPLLACHASDSEQPMLLFPKTALGTLKTVLLRTRDNKTTPGGSQKGALA
LSTQNLVMVAAQISRGMYHLTKKGLTHRDLAARNIYIHQNLHVRIGDRGLSWDLYPEEYSQMPDGEMCPVKWMAAEVLTD
RNYSHYSDVWSFGVVLWEIMTLGKIPYSENTAEEMVALLTAGQRLSQPKNCPDDLFVLMGWCWALTPTDRPRFSHLTLRL
KEFHEKICSFV
```

### Ci|198430807 vs Dysidea avara TR158463_c0_g1_i1 transcribed RNA sequence
```{bash}
>Da|TR158463_c0_g1_i1
SLKPAVDGSLAVLTAELNKSVHSLFSMKDILVDKNRIQLGELIKEGTYGRMYEGYLANTEDDVEGSVSIIVKTVSDITP
DQVVASLVDGCLVLRHIQHRHILQALGVYSNEQEPVMVLLARTTFGTLKQFLDDIRLQGGPQSSYGKLLLTQDMVFISAQ
IARGMYHLSRKGGIIHKDLAARNIYIHENLHVKIGDRGLSWDFYPDDYQVTPNGEYVPVKWASVEVLEEQQYSMYSDVWA
FGVVLWEIMTVGLVPYDSVTPNELVDYLKSGQRLTRPTNCPDDMFSIMGQCWALTPTDRPKFSHLTVGLDHFHLRLNSYI
```

```{bash}
touch my_data/Ci_198430807_vs_sponges.fasta

hmmscan --cut_ga --domtblout my_data/hmm_search/WIF_profile_Ci_198430807vssponges_domtbl.out my_data/pfam/Pfam-A.hmm my_data/Ci_198430807_vs_sponges.fasta > /dev/null
```


## Import hmmscan of WIF1 vs sponges results into R
```{r}
second_line_fields <- c("target_name", "accession_1", "tlen", "query_name",
                        "accession_2", "qlen",   "E-value_full_seq", "score_full_seq",
                        "bias_full_seq",
                        "this_domain_#",  "of_this_domain",  "c-Evalue_this_domain",
                        "i-Evalue_this_domain",  "score_this_domain", "bias_this_domain", 
                        "from_hmm_coord", "to_hmm_coord", "from_ali_coord", "to_ali_coord", 
                        "from_env_coord", "to_env_coord", "acc", "description_of_target_1",
                        "description_of_target_2","description_of_target_3")

WIF_hmmscan <- data.table::fread(cmd = "grep -v '^#' hmm_search/WIF_profile_Ci_198430807vssponges_domtbl.out | awk -v OFS='\t'  '{for(i=1;i<=29;i++) printf $i\"\t\"; print \"\"}'",  header = FALSE, sep = "\t",  fill=TRUE)

names(WIF_hmmscan)[1:25] <- second_line_fields

WIF_hmmscan %>% 
  distinct(target_name, query_name) %>% 
  pivot_wider(names_from = target_name, values_from = target_name,values_fill = "0") %>% 
  unite(col = "domains",2:11) %>% 
  mutate(domains = domains %>% str_remove_all("0_") %>% str_remove_all("_0"),
         query_name = query_name %>% str_replace("\\|","_")) %>% 
  unite(query_name, 1:2, sep = "__DOMAIN__") %>% view()
```
